kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/PaginatedScanFilesIteratorImpl.java

```diff
@@ -34,21 +46,68 @@ public PaginatedScanFilesIteratorImpl(
     this.pageSize = paginationContext.getPageSize();
   }
 
+  // Bug: we can't call hasNext() twice
   @Override
-  public Row getCurrentPageToken() {
-    return null;
+  public boolean hasNext() {
+    if (closed) {
+      throw new IllegalStateException("Can't call `hasNext` on a closed iterator.");
+    }
+    if (!nextBatch.isPresent()) {
+      prepareNext();
+    }
+    return nextBatch.isPresent();
   }
 
-  @Override
-  public boolean hasNext() {
-    return false;
+  private void prepareNext() {
+    if (numScanFilesReturned >= pageSize) {
+      return;
+    }
+    if (originalIterator.hasNext()) {
+      FilteredColumnarBatch batch = originalIterator.next();
+      String fileName = batch.getFilePath().get();
```

Comment:
how do we _know_ it has a file name? please check if -- we certainly **expect** and **require** that it does, so check if it does not and then throw an exception (IllegalStateException)